{% extends 'base.html' %}

{% block title %}Instructions{% endblock %}

{% block content %}
  <section class="py-5">
    <div class="container col-12 col-lg-9">
      <h2 class="mb-4 text-center fw-normal">Instructions</h2>
      <hr>

      <div class="col-12 col-lg-9 bg-light p-4 mt-3 mx-auto">
        <h5>
          <img src="static/images/info-square.svg" class="me-1 mb-1" width="20" height="20" alt="info-square">
          Two ways to use
          <b>sfkit</b>
        </h5>

        <p>In an
          <b>auto-configured</b>
          mode, once the study is configured and launched on the website, sfkit will automatically create the computing environment and deploy the joint analysis protocol in the Google Cloud Platform (GCP). To allow this automation, sfkit will ask for a minimal set of permissions for your GCP project.</p>

        <p>In a
          <b>user-configured</b>
          mode, once the study is configured, you can easily run the protocol on your own machine (and your collaborators on their machines) using the sfkit command-line interface (CLI).
        </p>
      </div>

      <div class="row">
        <h4 class="my-4 fw-normal">Prerequisites</h4>
        <p>
          To run a study using
          <b>sfkit</b>, you will need either:
        </p>
        <div>
          <ol>
            <li>
              A Google Cloud Platform (GCP) account that can create and manage virtual machines (VMs) in the cloud.
            </li>
            <li>
              A machine of your own with a nework connection, if you are running a user-configured study.
            </li>
          </ol>

          <p>See
            <i>Machine Recommendations</i>
            on the bottom of this page for guidance on machine types.
          </p>

          <span class="text-muted">
            Note: For the
            <a href="{{ url_for('general.tutorial') }}" class="text-decoration-none">Tutorial</a>, you can use our GCP project and example data for testing purposes.
          </span>
        </div>
      </div>

      <div>
        <h4 class="my-4 fw-normal">Data Preparation</h4>
        <div class="row p-4 bg-light rounded">
          <ul class="nav nav-tabs" role="tablist">
            <li class="nav-item">
              <button class="nav-link active" data-bs-toggle="tab" data-bs-target="#data-mpcgwas" type="button">MPC-GWAS workflow</button>
            </li>
            <li class="nav-item">
              <button class="nav-link" data-bs-toggle="tab" data-bs-target="#data-sfgwas" type="button">SF-GWAS workflow</button>
            </li>
            <li class="nav-item">
              <button class="nav-link" data-bs-toggle="tab" data-bs-target="#data-sfpca" type="button">SF-PCA workflow</button>
            </li>
          </ul>
          <div class="row tab-content">
            <div class="container tab-pane fade show active" id="data-mpcgwas">
              <div class="row mt-3">
                <p>
                  For GWAS, each user holds a portion of the horizontally distributed input genotype matrix (in which each row is a data sample and the features correspond to SNPs), covariate matrix and phenotype vector. These users' local files must be named and formatted as follows:
                </p>
                <div>
                  <ul>
                    <li>
                      <code>geno.txt</code>
                      - The genotype matrix, or minor allele dosage matrix, is stored as a tab-separated file in which the SNP values (i.e., features) are encoded as genotype scores (i.e., as 0, 1, or 2).
                    </li>
                    <li>
                      <code>pos.txt</code>
                      - This file must accompany the genotype matrix and stores the genomic positions of the SNPs in a 2-columns file where each row contains the chromosome number and the position in the chromosome of the corresponding SNP, separated by a tabulation.
                    </li>
                    <li>
                      <code>cov.txt</code>
                      - A tab-separated file storing the covariate matrix in which each row is a sample, and each column is a covariate, e.g., patient older than 50 years old. We assume all covariates are binary.
                    </li>
                    <li>
                      <code>pheno.txt</code>
                      - The phenotype vector, e.g., containing the infection status of each patient, is stored in a single-column file.
                    </li>
                  </ul>
                </div>
              </div>
            </div>

            <div class="container tab-pane fade" id="data-sfgwas">
              <div class="row mt-3">
                <p>
                  For this workflow, the input data are similar to those in the previous workflow, except for the following differences:
                </p>
                <div>
                  <ul>
                    <li>
                      <code>geno/chr[1-22].[pgen|psam|pvar]:</code>
                      The genotype or minor allele dosage matrix is encoded using the PGEN file format for each chromosome. This format has been introduced in the standard
                      <a class="text-decoration-none" href="https://www.cog-genomics.org/plink/2.0/formats">PLINK2</a>
                      tool for genomic data processing as an efficient way to store large-scale genomic datasets. Note that this file encodes the genomic positions of the SNPs directly.
                    </li>
                    <li>
                      <code>sample_keep.txt:</code>
                      This file accompanies the genotype matrix and lists the sample IDs from the .psam file to include in the analysis. This file is required to comply with the standard format proposed in PLINK2 (see the --keep option in the PLINK2 documentation).
                    </li>
                    <li>
                      <code>pheno.txt:</code>
                      As before, each line includes the phenotype under study for each sample.
                    </li>
                    <li>
                      <code>cov.txt:</code>
                      Each line includes a tab-separated list of covariates for each sample. Unlike in the previous workflow, the covariates and phenotypes in this workflow are not required to be binary.
                    </li>
                  </ul>
                </div>
              </div>
            </div>

            <div class="container tab-pane fade" id="data-sfpca">
              <div class="row mt-3">
                <p>
                  Each user data, consisting in an horizontal partition of the input matrix (in which the rows are the data samples and the columns correspond to the features) must be locally stored as a single tab-separated file called
                  <code>data.txt</code>.
                </p>
              </div>
            </div>

          </div>
        </div>
      </div>

      <div>
        <h4 class="my-4 fw-normal">Getting Started</h4>
        <p>
          When you are ready to run a study, go to
          <a href="{{ url_for('studies.index') }}" class="text-decoration-none">Studies</a>
          to create or join a study.
        </p>

        <div>
          <p>If you are creating a study, you will need to:</p>
          <ul>
            <li>
              Choose one of our
              <a href="{{ url_for('general.workflows') }}" class="text-decoration-none">Workflows</a>
              for the study.
            </li>
            <li>
              Specify whether you want the study to be auto-configured or user-configured (described below).
            </li>
            <li>
              Edit your study parameters.
            </li>
          </ul>
        </div>
        <p>
          If you are joining a study, you will only need to review and update the study parameters to provide information about your dataset.
        </p>
      </div>

      <div>
        <h4 class="my-4 fw-normal">Configuration Options</h4>
        <div class="row p-4 bg-light rounded">
          <ul class="nav nav-tabs" role="tablist">
            <li class="nav-item">
              <button class="nav-link active" data-bs-toggle="tab" data-bs-target="#Auto-configured" type="button">Auto-configured</button>
            </li>
            <li class="nav-item">
              <button class="nav-link" data-bs-toggle="tab" data-bs-target="#User-configured" type="button">User-configured</button>
            </li>
          </ul>
          <div class="row tab-content">
            <div class="container tab-pane fade show active" id="Auto-configured">
              <div class="row mt-3">
                <p>
                  The sfkit portal will set up the machine and run the study for you. This option is currently supported using the
                  <a class="text-decoration-none" href="https://cloud.google.com/" target="_blank">Google Cloud Platform (GCP)</a>. You will need to give sfkit limited permissions to interact with your GCP project. You will be walked through the specifics of this process once you create a study with this option.
                  <!-- If you are new to GCP, you can find lots of information and tutorials on <a class="text-decoration-none" href="https://cloud.google.com/">their website</a>. -->
                </p>
                <p>This is recommended for users who are not familiar with the command line and/or want to get started quickly.</p>
              </div>
            </div>
            <div class="container tab-pane fade" id="User-configured">
              <div class="row mt-3">
                <p>
                  The
                  <a class="text-decoration-none" href="https://sfkit.readthedocs.io/en/latest/tutorial.html" target="_blank">sfkit Command-Line Interface</a>
                  will walk you through each step of the workflow so you can run the study on your own machine.
                </p>
                <p>This is recommended for users who are familiar with the command line and want to directly control their computing environment.</p>
              </div>
            </div>
          </div>
        </div>
      </div>

      <div class="mt-4">
        <p>For both configuration options, the following steps will be run on your machine:</p>
      </div>

      <div>

        <h4 class="my-4 fw-normal">sfkit Process</h4>

        <div class="row justify-content-around">
          <div class="col-6 col-md-2 my-3 btn d-flex align-items-stretch" id="a">
            <div class="position-relative p-4 text-center bg-light rounded">
              <span class="d-flex justify-content-center align-items-center position-absolute top-0 start-50 translate-middle rounded-circle bg-primary text-white" style="width: 32px;height: 32px;font-size: 0.8rem;">1</span>
              <h6 class="fw-bold mb-2">Authentication</h6>
              <img class="img-fluid" src="static/images/sfkit/auth.png" alt="">
            </div>
          </div>
          <div class="col-6 col-md-2 my-3 btn d-flex align-items-stretch" id="b">
            <div class="position-relative p-4 text-center bg-light rounded">
              <span class="d-flex justify-content-center align-items-center position-absolute top-0 start-50 translate-middle rounded-circle bg-primary text-white" style="width: 32px;height: 32px;font-size: 0.8rem;">2</span>
              <h6 class="fw-bold mb-2">Networking</h6>
              <img class="img-fluid" src="static/images/sfkit/networking.png" alt="">
            </div>
          </div>
          <div class="col-6 col-md-2 my-3 btn d-flex align-items-stretch" id="c">
            <div class="position-relative p-4 text-center bg-light rounded">
              <span class="d-flex justify-content-center align-items-center position-absolute top-0 start-50 translate-middle rounded-circle bg-primary text-white" style="width: 32px;height: 32px;font-size: 0.8rem;">3</span>
              <h6 class="fw-bold mb-2">Key Exchange</h6>
              <img class="img-fluid" src="static/images/sfkit/keys.png" alt="">
            </div>
          </div>
          <div class="col-6 col-md-2 my-3 btn d-flex align-items-stretch" id="d">
            <div class="position-relative p-4 text-center bg-light rounded">
              <span class="d-flex justify-content-center align-items-center position-absolute top-0 start-50 translate-middle rounded-circle bg-primary text-white" style="width: 32px;height: 32px;font-size: 0.8rem;">4</span>
              <h6 class="fw-bold mb-2">Data Validation</h6>
              <img class="img-fluid" src="static/images/sfkit/data.png" alt="">
            </div>
          </div>
          <div class="col-6 col-md-2 my-3 btn d-flex align-items-stretch" id="e">
            <div class="position-relative p-4 text-center bg-light rounded">
              <span class="d-flex justify-content-center align-items-center position-absolute top-0 start-50 translate-middle rounded-circle bg-primary text-white" style="width: 32px;height: 32px;font-size: 0.8rem;">5</span>
              <h6 class="fw-bold mb-2">Run Protocol</h6>
              <img class="img-fluid" src="static/images/sfkit/run_protocol.png" alt="">
            </div>
          </div>
        </div>

        <div class="row" id="Authentication" style="display: none">
          <div class="card card-body">
            <p>
              1.
              <b>Authentication</b>: Authenticate your machine with sfkit. This allows sfkit to ensure that the machine running the protocol is indeed yours. This is achieved by validating a secure key that is copied from the website to the machine.
            </p>
          </div>
        </div>
        <div class="row" id="Networking" style="display: none">
          <div class="card card-body">
            <p>
              2.
              <b>Networking</b>: Setup the networking for the machine. If you are running a user-configured study, you will be asked for the designated ports to use for the secure connections used by the protocol. Then, your IP address and ports will be shared with your collaborators via the sfkit portal.
            </p>
          </div>
        </div>
        <div class="row" id="KeyExchange" style="display: none">
          <div class="card card-body">
            <p>
              3.
              <b>Key Exchange</b>: Cryptographically secure private and public keys will be created (locally) for you. The public key will be shared with your collaborators via the sfkit portal. The private key will remain on your machine and will be used during the study protocol to establish secure connections between machines. See
              <a href="https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange" class="text-decoration-none">Diffie-Hellman Key Exchange</a>
              for more information.
            </p>
          </div>
        </div>
        <div class="row" id="DataValidation" style="display: none">
          <div class="card card-body">
            <p>
              4.
              <b>Data Validation</b>: Validate that the data you provided is in the correct format and is consistent with the study parameters you provided. This step does NOT communicate any of the data externally and only reports the validation result back to the sfkit portal to assess when the study is ready to be launched.
            </p>
          </div>
        </div>
        <div class="row" id="RunProtocol" style="display: none">
          <div class="card card-body">
            <p>
              5.
              <b>Run Protocol</b>: Run the collaborative analysis protocol that you have chosen. Your machine will use the ports and IP addresses provided in Step 2 to securely communicate with collaborators' machines. This step may take several hours depending on the study workflow and the dataset sizes. The results of the protocol can be conveniently accessed in the output location or via the portal if you have given sfkit permission to fetch results for you.
            </p>
          </div>
        </div>
      </div>

      <br>

      <section id="machine_recommendations">
        <h4 class="my-4 fw-normal">Machine Recommendations</h4>
        <p>
          If you are unsure what machine size or type to use for your study, you can use this tool to see our recommendation. Note that this guidance is based on the machines that are available on the Google Cloud Platform, but equivalent machines can be used instead.
        </p>
        <div class="row p-4 bg-light rounded">
          <ul class="nav nav-tabs" role="tablist">
            <li class="nav-item">
              <button class="nav-link active" data-bs-toggle="tab" data-bs-target="#MPCGWASmachine" type="button">
                MPC-GWAS workflow
              </button>
            </li>
            <li class="nav-item">
              <button class="nav-link" data-bs-toggle="tab" data-bs-target="#SFGWASmachine" type="button">
                SF-GWAS workflow
              </button>
            </li>
            <li class="nav-item">
              <button class="nav-link" data-bs-toggle="tab" data-bs-target="#SFPCAmachine" type="button">
                SF-PCA workflow
              </button>
            </li>
          </ul>

          <div class="tab-content" id="myTabContent">

            <div class="tab-pane fade show active" id="MPCGWASmachine" role="tabpanel">
              <div class="row mt-3">
                <div class="col-6">
                  <select class="form-select" id="num_rows_mpcgwas">
                    <option selected="selected">Approximately how many samples/individuals are in your dataset?</option>
                    <option value="1">fewer than 30,000</option>
                    <option value="2">30,000 - 100,000</option>
                    <option value="3">100,000 - 500,000</option>
                    <option value="4">more than 500,000</option>
                  </select>
                </div>
                <div class="col-6">
                  <select class="form-select" id="num_cols_mpcgwas">
                    <option selected="selected">Approximately how many SNPs are in your dataset?</option>
                    <option value="1">fewer than 700,000</option>
                    <option value="2">700,000 - 10,000,000</option>
                    <option value="3">10,000,000 - 30,000,000</option>
                    <option value="4">more than 30,000,000</option>
                  </select>
                </div>
                <div class="col mt-3">
                  <div class="card card-body" id="recommendation_mpcgwas">The default recommendation is an e2-highmem-16 machine with 128GB RAM and a boot disk of at least 128GB (estimated cost of $0.75/hour).</div>
                </div>
              </div>
            </div>

            <div class="tab-pane fade" id="SFGWASmachine" role="tabpanel">
              <div class="row mt-3">
                <div class="col-6">
                  <select class="form-select" id="num_rows_sfgwas">
                    <option selected="selected">Approximately how many samples/individuals are in your dataset?</option>
                    <option value="1">fewer than 30,000</option>
                    <option value="2">30,000 - 100,000</option>
                    <option value="3">100,000 - 500,000</option>
                    <option value="4">more than 500,000</option>
                  </select>
                </div>
                <div class="col-6">
                  <select class="form-select" id="num_cols_sfgwas">
                    <option selected="selected">Approximately how many SNPs are in your dataset?</option>
                    <option value="1">fewer than 700,000</option>
                    <option value="2">700,000 - 10,000,000</option>
                    <option value="3">10,000,000 - 30,000,000</option>
                    <option value="4">more than 30,000,000</option>
                  </select>
                </div>
                <div class="col mt-3">
                  <div class="card card-body" id="recommendation_sfgwas">The default recommendation is an e2-highmem-16 machine with 128GB RAM and a boot disk of at least 128GB (estimated cost of $0.75/hour).</div>
                </div>
              </div>
            </div>

            <div class="tab-pane fade" id="SFPCAmachine" role="tabpanel">
              <div class="row mt-3">
                <div class="col-6">
                  <select class="form-select" id="num_rows_sfpca">
                    <option selected="selected">Approximately how many rows are in your dataset?</option>
                    <option value="1">fewer than 30,000</option>
                    <option value="2">30,000 - 100,000</option>
                    <option value="3">100,000 - 500,000</option>
                    <option value="4">more than 500,000</option>
                  </select>
                </div>
                <div class="col-6">
                  <select class="form-select" id="num_cols_sfpca">
                    <option selected="selected">Approximately how many columns are in your dataset?</option>
                    <option value="1">fewer than 700,000</option>
                    <option value="2">700,000 - 10,000,000</option>
                    <option value="3">10,000,000 - 30,000,000</option>
                    <option value="4">more than 30,000,000</option>
                  </select>
                </div>
                <div class="col mt-3">
                  <div class="card card-body" id="recommendation_sfpca">The default recommendation is an e2-highmem-16 machine with 128GB RAM and a boot disk of at least 128GB (estimated cost of $0.75/hour).</div>
                </div>
              </div>
            </div>

          </div>
        </div>
      </section>

    </div>
  </section>
{% endblock %}

{% block javascript %}
  <script>
    function recommendMachine(workflow) {
      var base = "Based on your selection, we recommend you use an ";
      var recommendations = {
        1: 'e2-highmem-16 with at least a 128 GB boot disk (estimated cost of $0.75/hour).',
        2: 'n2-highmem-32 with at least a 256 GB boot disk (estimated cost of $1.75/hour).',
        3: 'n2-highmem-64 with at least a 512 GB boot disk (estimated cost of $3.50/hour).',
        4: 'n2-highmem-128 with at least a 1024 GB boot disk (estimated cost of $6.50/hour).'
      }
      var num_rows = parseInt($('#num_rows_' + workflow).val()) || 0;
      var num_cols = parseInt($('#num_cols_' + workflow).val()) || 0;
      var selection = Math.max(num_rows, num_cols, 1);
      $('#recommendation_' + workflow).text(base + recommendations[selection]);
    }

    const workflows = ['sfgwas', 'sfpca', 'mpcgwas'];
    workflows.forEach(function (workflow) {
      $('#num_rows_' + workflow).change(function () {
        recommendMachine(workflow);
      });
      $('#num_cols_' + workflow).change(function () {
        recommendMachine(workflow);
      });
    });
  </script>

  <script type="module" src="{{ url_for('static', filename='javascript/sfkit_explanations.js') }}"></script>
{% endblock %}